### ISSUE TITLE
Investigate on codec reconfiguration for a active connection

### ISSUE THREAD

**Issue by mp911de** [lettuce-io/lettuce-core#632]
Using different codecs requires to run multiple connections as each connection is typed with a specific codec. We should investigate whether it's possible to retype a connection to create a new connection facade that provides its functionality with a different codec:

```
StatefulRedisConnection<String, String> connection  = …

StatefulRedisConnection<byte[], byte[]> binary = connection.withCodec(ByteArrayCodec.INSTANCE);
```

**Comment by TheSpiritXIII** [comment]
I agree with this suggestion. My use case is this: I want to reuse a connection pool but serialize differently depending on the data structure, e.g. single Redis database but with maps and lists used for different purposes.

While I agree this depends largely on the serialization strategy (deserializing between different key types may be tricky and lead to ill effects), the user should decide whether they want to. In most cases, e.g. when my serialization strategy is JSON/XML/etc, it is valid.

I could do this in Spring Data Redis via `RedisTemplate<K, V>`. There, I could have different templates with different serialization strategies utilizing the same connection factory. We could look through what Spring has done, because they have a slightly better API for this (however, I opted to use Lettuce because they have an asynchronous API, while Spring does not).

**Comment by mp911de** [comment]
Looking at the API, we could introduce `StatefulRedisConnection.withCodec(…)` to create a new object that exposes sync/async/reactive API objects that are associated with the new `RedisCodec`. The thing here is that `StatefulRedisConnection` isn't the right type to return as we would have two connection object instances whereas we have only a single connection. Finding a good name for the type seems the actual challenge here.

@TheSpiritXIII You could create `RedisAsyncCommandsImpl` instances passing in the connection object and a `RedisCodec` to create multiple async API facades for a single connection.

Another caveat is that `RedisAsyncCommandsImpl` and other command API types return `StatefulRedisConnection`.

**Comment by osoriano** [comment]
I'm migrating from Jedis and hitting this issue as well. Jedis [supports byte[]/String methods out of the box](https://github.com/xetorthio/jedis/wiki/AdvancedUsage#a-note-about-string-and-binary---what-is-native), but for Lettuce, I'm planning to use a `RedisCodec<ByteBuffer,ByteBuffer>` and manually encode/decode the Lettuce calls.

**Comment by mp911de** [comment]
Lettuce supports binary and `String` types out of the box (`ByteArrayCodec.INSTANCE`, `StringCodec.UTF8`, `StringCodec.ASCII`).

**Comment by little-fish** [comment]
Hello.
Is there any progress related to this feature?
Thank you.

**Comment by mp911de** [comment]
Right now, there's no proper way how to achieve this functionality.

We either get hold of another connection object that doesn't represent the actual connection or we get a back-reference to a connection that isn't using the reconfigured codec (`RedisCommands.getStatefulConnection()`).

That being said, if you have another proposal, we're happy to discuss it here.

**Comment by tishun** [comment]
Hey @mgroth0 ,

thanks for reaching out, please allow me to confirm my understanding of the issue.

> But then I ran into an issue because the command `KEYS` takes a glob pattern, not a regular key, which isn't compatible with `MyRedisKey`. 

So basically you'd like to call `KEYS` with a GLOB pattern, but if you configure the driver to use a custom codec that treats K as `MyRedisKey` you could not. What prevents you form adding this logic in the `MyRedisKey`? E.g. having an implementation that is not a specific key, but a GlobMyRredisKey and then the codec would handle that depending on the implementation used?

**Comment by tishun** [comment]
To be fair, specifically about the signature of the keys() method, I gave some reservations that it is following the right now.

Every redis command has a specified [request / response policy](https://redis.io/docs/latest/develop/reference/command-tips/). In the case of the `KEYS` command we have a `request_policy:all_shards` :

```
127.0.0.1:6379> command info keys
    1)  1) "keys"
    2) (integer) 2
    3) 1) readonly
    4) (integer) 0
    5) (integer) 0
    6) (integer) 0
    7) 1) @keyspace
       2) @read
       3) @slow
       4) @dangerous
    8) 1) "request_policy:all_shards"
       2) "nondeterministic_output_order"
    9) (empty array)
   10) (empty array)
```

This means that the parameter should NOT be treated as a key at all - in a cluster environment it's hash should not be used to determine the node that would handle the request, but instead it needs to be treated as a keyless command - sending the request to all nodes and aggregating the results locally.

So - for this specific case - I would say we need to make a breaking change and change the API of the driver.

Added a task for that in #3311

**Comment by tishun** [comment]
> At the time that I wrote the above comment, I only was thinking about key globs for KEYS. I have since realized that I also needed to have similar workarounds for hash fields and channel names which also use the K from the key type

I am a bit confused. Why is it a problem for hash fields and channels?
